lib/ts/PriorityQueue.h

```diff
@@ -0,0 +1,207 @@
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#ifndef __PRIORITY_QUEUE_H__
+#define __PRIORITY_QUEUE_H__
+
+#include "ts/Diags.h"
+#include "ts/Vec.h"
+
+template <typename T> struct PriorityQueueEntry {
+  PriorityQueueEntry(T n) : index(0), node(n) {}
+
+  uint32_t index;
+  T node;
+};
+
+template <typename T> class PriorityQueue
+{
+public:
+  PriorityQueue() {}
+  ~PriorityQueue() {}
+
+  const bool empty();
+  PriorityQueueEntry<T> *top();
+  void pop();
+  void push(PriorityQueueEntry<T> *);
+  void update(PriorityQueueEntry<T> *);
+  void update(PriorityQueueEntry<T> *, bool);
+  const Vec<PriorityQueueEntry<T> *> &dump() const;
+
+private:
+  Vec<PriorityQueueEntry<T> *> _v;
+
+  void _swap(uint32_t, uint32_t);
+  void _bubble_up(uint32_t);
+  void _bubble_down(uint32_t);
+};
+
+template <typename T>
+const Vec<PriorityQueueEntry<T> *> &
+PriorityQueue<T>::dump() const
+{
+  return _v;
+}
+
+template <typename T>
+const bool
+PriorityQueue<T>::empty()
+{
+  return _v.length() == 0;
+}
+
+template <typename T>
+void
+PriorityQueue<T>::push(PriorityQueueEntry<T> *entry)
+{
+  ink_release_assert(entry != NULL);
+
+  int len = _v.length();
+  _v.push_back(entry);
+  entry->index = len;
+
+  _bubble_up(len);
+}
+
+template <typename T>
+PriorityQueueEntry<T> *
+PriorityQueue<T>::top()
+{
+  if (empty()) {
+    return NULL;
+  } else {
+    return _v[0];
+  }
+}
+
+template <typename T>
+void
+PriorityQueue<T>::pop()
+{
+  if (empty()) {
+    return;
+  }
+
+  _v[0] = _v[_v.length() - 1];
+  _v.pop();
+  _bubble_down(0);
+}
+
+template <typename T>
+void
+PriorityQueue<T>::update(PriorityQueueEntry<T> *entry)
+{
+  ink_release_assert(entry != NULL);
+
+  if (empty()) {
+    return;
+  }
+
+  // One of them will works whether weight is increased or decreased
+  _bubble_down(entry->index);
+  _bubble_up(entry->index);
+}
+
+template <typename T>
+void
+PriorityQueue<T>::update(PriorityQueueEntry<T> *entry, bool increased)
+{
+  ink_release_assert(entry != NULL);
+
+  if (empty()) {
+    return;
+  }
+
+  if (increased) {
+    _bubble_down(entry->index);
+  } else {
+    _bubble_up(entry->index);
+  }
+}
+
+template <typename T>
+void
+PriorityQueue<T>::_swap(uint32_t i, uint32_t j)
+{
+  PriorityQueueEntry<T> *tmp = _v[i];
+  _v[i] = _v[j];
+  _v[j] = tmp;
+
+  _v[i]->index = i;
+  _v[j]->index = j;
+}
+
+
+template <typename T>
+void
+PriorityQueue<T>::_bubble_up(uint32_t index)
+{
+  if (empty()) {
+    ink_release_assert(false);
+  }
+
+  uint32_t parent;
+  while (index != 0) {
+    parent = (index - 1) / 2;
+    if (*(_v[index]->node) < *(_v[parent]->node)) {
```

Comment:
It might be better to pass a comparator to the constructor. ``<`` may be overloaded for other purposes.